File: Data Structures in C/Linked Lists/Doubly Linked List/DoublyLinkedListInsertion.c
#include<stdio.h>
#include<stdlib.h>
typedef struct Node
{
    int data;
    struct Node *next;
    struct Node *prev;
}Node;
Node* insert(Node *head,int data,int pos)
{
    Node *n=(Node *)malloc(sizeof(Node));
    n->data=data;
    n->next=NULL;
    n->prev=NULL;
    Node *h=head;
    if(pos==1 || head==NULL)
    {
        n->next=h;
        if(head!=NULL)
        h->prev=n;
        head=n;
    }
    else
    {
        int k=1;
        while(h->next!=NULL && k!=pos-1){
            h=h->next;
            k++;
        }
        if(h->next==NULL)
        {
            h->next=n;
            n->prev=h;
        }
        else{
        n->next=h->next;
        h->next->prev=n;
        h->next=n;
        n->prev=h;
        }
    }
    return head;
}
void traverse(Node *head)
{
    Node *p;
    while(head!=NULL)
    {
        printf("%d ",head->data);
        head=head->next;
        if(head!=NULL)
        p=head;
    }
    printf("\nThe elements in reverse order :\n");
    while(p!=NULL)
    {
        printf("%d ",p->data);
        p=p->prev;
    }
    printf("\n");
}
int main()
{
    printf("Enter the number of elements you want to enter :\n");
    int num;
    scanf("%d",&num);
    Node *head=NULL;
    Node *p=NULL;
    Node *n=NULL;
    printf("Enter the elements :\n");
    for(int i=0;i<num;i++)
    {
        if(head==NULL)
        {
            p=(Node *)malloc(sizeof(Node));
            scanf("%d",&p->data);
            p->prev=NULL;
            p->next=NULL;
            head=p;
        }
        else
        {
            n=(Node *)malloc(sizeof(Node));
            scanf("%d",&n->data);
            n->next=NULL;
            n->prev=p;
            p->next=n;
            p=n;
        }
        
    }
    printf("The elements are :\n");
    traverse(head);
    printf("Enter the data you want to enter and it's position :\n");
    int d,pos;
    scanf("%d%d",&d,&pos);
    head=insert(head,d,pos);
    printf("After insertion new Linked List is :\n");
    traverse(head);
    return 0;
}

File: Data Structures in JAVA/Linked Lists/Doubly Linked List/src/DoublyLinkedListInsertion.java
/*
   This program inserts a node at any given position in a Doubly Linked List in JAVA
   Author - Rajarshi Sengupta
   Github - https://github.com/rajarshisg
   Date - 20/08/2020 (dd/mm/yyyy)
*/
import java.util.Scanner;
public class DoublyLinkedListInsertion {
    Node insert(Node head,int data,int pos)
    {
    	Node n=new Node(data);
    	Node h=head;
    	if(pos==1 || head==null)
    	{
    		n.next=head;
    		n.prev=null;
    		head.prev=n;
    		head=n;
    		
    	}
    	else
    	{
    		int k=1;
    		while(k!=pos-1 && h.next!=null) {
    			h=h.next;
    			k++;
    		}
    		
    		n.next=h.next;
    		if(h.next!=null)
    		h.next.prev=n;
    		h.next=n;
    		n.prev=h;
    	}
    	return head;
 }
    void traverse(Node head) {
       System.out.println("The Doubly Linked List :");
  	   Node h=null;
  	   while(head!=null)
  	   {
  		   h=head;
  		   System.out.print(head.data+" ");
  		   head=head.next;
  	   }
  	   System.out.println();
  	   System.out.println("The Doubly Linked List in reverse order :");
  	   while(h!=null) {
  		   System.out.print(h.data+" ");
  		   h=h.prev;
  	   }
  	   System.out.println();
     }
   public static void main(String args[]) {
      Scanner sc=new Scanner(System.in);
      System.out.println("Enter the number of nodes you want :");
      int num=sc.nextInt();
      Node<Integer> head=null,n=null,m=null;
      System.out.println("Enter the elements :");
      for(int i=1;i<=num;i++)
      {
    	  if(head==null) {
    		n=new Node<Integer>(sc.nextInt());
    	    head=n;
    	}
        else{
    	    m=new Node<Integer>(sc.nextInt());
    	    n.next=m;
    	    m.prev=n;
    	    n=m;
    	 }
      }
     DoublyLinkedListInsertion ob=new DoublyLinkedListInsertion();
     ob.traverse(head);  
     System.out.println("Enter the data and the position you want to enter it at :");
     int data=sc.nextInt();
     int pos=sc.nextInt();
     head=ob.insert(head,data,pos);
     ob.traverse(head);
    }

}
